gapidapk/gapidapk.go

```diff
@@ -207,3 +220,124 @@ func LayerName(vulkan bool) string {
 		return LibGAPIIName
 	}
 }
+
+func ensurePerfettoProducerLaunched(ctx context.Context, d adb.Device, useNative bool) error {
+	// Always kill the exisiting perfetto producer launcher, otherwise perfetto client
+	// may end up quering back wrong data source information.
+	d.Shell("killall", perfettoProducerLauncher).Run(ctx)
+	hasNativeSupport, err := d.HasNativeGpuProfilingSupport(ctx)
+	if err != nil {
+		return log.Err(ctx, err, "Fail to query native GPU profiling support.")
+	}
+	if !hasNativeSupport && useNative {
+		log.E(ctx, "Native producers are requested but no native GPU profiling support found.")
+		return log.Err(ctx, nil, "No native GPU profiling support found.")
+	}
+	startSignal, startFunc := task.NewSignal()
+	startFunc = task.Once(startFunc)
+	crash.Go(func() {
+		err := launchPerfettoProducer(ctx, d, useNative, startFunc)
+		if err != nil {
+			log.E(ctx, "[ensurePerfettoProducerLaunched] error: %v", err)
+		}
+
+		// Ensure the start signal is fired on failure/immediate return.
+		startFunc(ctx)
+	})
+	startSignal.Wait(ctx)
+	// TODO(lpy): Data sources are queried before gpu.counters is registered.
+	time.Sleep(1 * time.Second)
```

Comment:
That looks a bit hacky, but I guess there is no better way for now?
Maybe track with a proper bug ID in the TODO?